ts/nni_manager/core/tuner_command_channel/websocket_channel.ts

```diff
@@ -0,0 +1,146 @@
+    return channelSingleton;
+}
+
+/**
+ *  The callback to serve WebSocket connection request. Used by REST server module.
+ *  It should only be invoked once, or an error will be raised.
+ *
+ *  Type hint of express-ws is somewhat problematic. Don't want to waste time on it so use `any`.
+ **/
+export function serveWebSocket(ws: WebSocket): void {
+    channelSingleton.setWebSocket(ws);
+}
+
+class WebSocketChannelImpl {
+    private deferredInit: Deferred<void> = new Deferred<void>();
+    private emitter: EventEmitter = new EventEmitter();
+    private heartbeatTimer!: NodeJS.Timer;
+    private serving: boolean = false;
+    private waitingPong: boolean = false;
+    private ws!: WebSocket;
+
+    public setWebSocket(ws: WebSocket): void {
+        if (this.ws !== undefined) {
+            logger.error('A second client is trying to connect');
+            ws.close(4030, 'Already serving a tuner.');
+            return;
+        }
+
+        logger.debug('Connected.');
+        this.serving = true;
+
+        this.ws = ws;
+        ws.on('close', () => { this.handleError(new Error('tuner_command_channel: Tuner closed connection')); });
+        ws.on('error', this.handleError.bind(this));
+        ws.on('message', this.receive.bind(this));
+        ws.on('pong', () => { this.waitingPong = false; });
+
+        this.heartbeatTimer = setInterval(this.heartbeat.bind(this), heartbeatInterval);
+        this.deferredInit.resolve();
+    }
+
+    public init(): Promise<void> {
+        logger.debug(this.ws === undefined ? 'Waiting connection...' : 'Initialized.');
+        return this.deferredInit.promise;
+    }
+
+    public async shutdown(): Promise<void> {
+        if (this.ws === undefined) {
+            return;
+        }
+        clearInterval(this.heartbeatTimer);
+        this.serving = false;
+        this.emitter.removeAllListeners();
+    }
+
+    public sendCommand(command: string): void {
+        assert.ok(this.ws !== undefined);
+
+        logger.debug('Sending', command);
+        this.ws.send(command);
+
+        if (this.ws.bufferedAmount > command.length + 1000) {
+            logger.warning('Sending too fast! Try to reduce the frequency of intermediate results.');
+        }
+    }
+
+    public onCommand(callback: (command: string) => void): void {
+        this.emitter.on('command', callback);
+    }
+
+    public onError(callback: (error: Error) => void): void {
+        this.emitter.on('error', callback);
+    }
+
+    private heartbeat(): void {
+        if (this.waitingPong) {
+            this.ws.terminate();  // this will trigger "close" event
+            this.handleError(new Error('tuner_command_channel: Tuner loses responsive'));
+        }
+
+        this.waitingPong = true;
+        this.ws.ping();
+    }
+
+    private receive(data: Buffer, _isBinary: boolean): void {
+        logger.debug('Received', data);
+        this.emitter.emit('command', data.toString());
```

Comment:
`_isBinary` is not used?

Reply:
I can't understand its behavior. Sometimes it holds unexpected value. So I always call `toString()`.

Reply:
Better to add comments (maybe a SO link) about this.

Reply:
Oh, it's Python part's problem. According to the type hint node API always provides a `Buffer` instead of `string`, so `toString()` is required anyway.

Reply:
To my understanding, WebSocket protocol send a bunch of plain data and attach a header indicating whether its binary or text. Node's WebSocket implementation simply passes the data and header to users as is. So we don't need the header when reading message.